File: java/base/src/main/vsdk/toolkit/environment/geometry/geometricProcessing/polyhedralBoundedSolidOperators/_PolyhedralBoundedSolidFallbackGeometry.java
package vsdk.toolkit.environment.geometry.geometricProcessing.polyhedralBoundedSolidOperators;

import java.util.ArrayList;
import java.util.Collections;

import vsdk.toolkit.common.linealAlgebra.Vector3Dd;
import vsdk.toolkit.environment.geometry.volume.polyhedralBoundedSolid.PolyhedralBoundedSolid;
import vsdk.toolkit.environment.geometry.volume.polyhedralBoundedSolid.nodes._PolyhedralBoundedSolidFace;
import vsdk.toolkit.environment.geometry.volume.polyhedralBoundedSolid.nodes._PolyhedralBoundedSolidLoop;

/**
Low-level geometric primitives shared by the structural-shape boolean fallback
builders (profile-difference, axis-aligned cell, orthogonal-profile and
offset-cylinder families). Extracted from
{@link PolyhedralBoundedSolidSetOperator} in Stage 7 R2 so that more than one
fallback family can reuse the same coordinate/profile helpers without
duplicating them. Pure code motion; no behavior change.

<p>Extends {@link _PolyhedralBoundedSolidOperator} solely to inherit the shared
{@code numericContext} tolerance state used by the epsilon comparisons.</p>
 */
final class _PolyhedralBoundedSolidFallbackGeometry
    extends _PolyhedralBoundedSolidOperator
{
    static double coordinate(Vector3Dd p, int axis)
    {
        if ( axis == 0 ) {
            return p.x();
        }
        if ( axis == 1 ) {
            return p.y();
        }
        return p.z();
    }

    static boolean sameCoordinate(double a, double b)
    {
        return Math.abs(a - b) <= numericContext.bigEpsilon();
    }

    static boolean boundsMatch(double[] a, double[] b)
    {
        int i;

        if ( a == null || b == null || a.length < 6 || b.length < 6 ) {
            return false;
        }
        for ( i = 0; i < 6; i++ ) {
            if ( !sameCoordinate(a[i], b[i]) ) {
                return false;
            }
        }
        return true;
    }

    static void addUniqueCoordinate(ArrayList<Double> values,
                                            double value)
    {
        int i;

        for ( i = 0; i < values.size(); i++ ) {
            if ( sameCoordinate(values.get(i), value) ) {
                return;
            }
        }
        values.add(value);
        Collections.sort(values);
    }

    static ArrayList<Double> uniqueVertexCoordinates(
        PolyhedralBoundedSolid solid,
        int axis)
    {
        ArrayList<Double> values;
        int i;

        values = new ArrayList<Double>();
        for ( i = 0; i < solid.getVerticesList().size(); i++ ) {
            addUniqueCoordinate(values,
                coordinate(solid.getVerticesList().get(i).position, axis));
        }
        return values;
    }

    static double signedAreaOnYZ(ArrayList<Vector3Dd> profile)
    {
        double area;
        int i;

        area = 0.0;
        for ( i = 0; i < profile.size(); i++ ) {
            Vector3Dd a;
            Vector3Dd b;

            a = profile.get(i);
            b = profile.get((i + 1) % profile.size());
            area += a.y() * b.z() - b.y() * a.z();
        }
        return area * 0.5;
    }

    static ArrayList<Vector3Dd> extractProfileAtX(
        PolyhedralBoundedSolid solid,
        double x)
    {
        ArrayList<Vector3Dd> best;
        double bestArea;
        int i;

        best = null;
        bestArea = 0.0;
        for ( i = 0; i < solid.getPolygonsList().size(); i++ ) {
            _PolyhedralBoundedSolidFace face;
            int j;

            face = solid.getPolygonsList().get(i);
            for ( j = 0; j < face.boundariesList.size(); j++ ) {
                _PolyhedralBoundedSolidLoop loop;
                ArrayList<Vector3Dd> profile;
                double area;
                int k;
                boolean onPlane;

                loop = face.boundariesList.get(j);
                if ( loop.halfEdgesList.size() < 3 ) {
                    continue;
                }
                profile = new ArrayList<Vector3Dd>();
                onPlane = true;
                for ( k = 0; k < loop.halfEdgesList.size(); k++ ) {
                    Vector3Dd p;

                    p = loop.halfEdgesList.get(k).startingVertex.position;
                    if ( !sameCoordinate(p.x(), x) ) {
                        onPlane = false;
                        break;
                    }
                    profile.add(new Vector3Dd(p));
                }
                if ( !onPlane ) {
                    continue;
                }
                area = Math.abs(signedAreaOnYZ(profile));
                if ( area > bestArea ) {
                    bestArea = area;
                    best = profile;
                }
            }
        }
        return best;
    }

    static boolean sameProfilePoint(Vector3Dd a, Vector3Dd b)
    {
        return sameCoordinate(a.x(), b.x()) &&
               sameCoordinate(a.y(), b.y()) &&
               sameCoordinate(a.z(), b.z());
    }

    static void appendProfilePoint(ArrayList<Vector3Dd> profile,
                                           Vector3Dd point)
    {
        if ( !profile.isEmpty() &&
             sameProfilePoint(profile.get(profile.size() - 1), point) ) {
            return;
        }
        profile.add(point);
    }

    static Vector3Dd projectProfilePoint(Vector3Dd point,
                                                double x,
                                                double zCut)
    {
        double z;

        z = point.z();
        if ( sameCoordinate(z, zCut) ) {
            z = zCut;
        }
        return new Vector3Dd(x, point.y(), z);
    }

    static Vector3Dd intersectProfileSegmentAtZ(Vector3Dd a,
                                                       Vector3Dd b,
                                                       double x,
                                                       double zCut)
    {
        double t;
        double y;

        if ( sameCoordinate(a.z(), b.z()) ) {
            return new Vector3Dd(x, a.y(), zCut);
        }
        t = (zCut - a.z()) / (b.z() - a.z());
        y = a.y() + (b.y() - a.y()) * t;
        return new Vector3Dd(x, y, zCut);
    }

    static ArrayList<Vector3Dd> clipProfileAboveZ(
        ArrayList<Vector3Dd> profile,
        double x,
        double zCut)
    {
        ArrayList<Vector3Dd> clipped;
        Vector3Dd previous;
        boolean previousInside;
        int i;

        clipped = new ArrayList<Vector3Dd>();
        if ( profile == null || profile.size() < 3 ) {
            return clipped;
        }

        previous = profile.get(profile.size() - 1);
        previousInside = previous.z() + numericContext.bigEpsilon() >= zCut;
        for ( i = 0; i < profile.size(); i++ ) {
            Vector3Dd current;
            boolean currentInside;

            current = profile.get(i);
            currentInside = current.z() + numericContext.bigEpsilon() >= zCut;

            if ( currentInside ) {
                if ( !previousInside ) {
                    appendProfilePoint(clipped,
                        intersectProfileSegmentAtZ(previous, current, x, zCut));
                }
                appendProfilePoint(clipped,
                    projectProfilePoint(current, x, zCut));
            }
            else if ( previousInside ) {
                appendProfilePoint(clipped,
                    intersectProfileSegmentAtZ(previous, current, x, zCut));
            }

            previous = current;
            previousInside = currentInside;
        }

        if ( clipped.size() > 1 &&
             sameProfilePoint(clipped.get(0),
                 clipped.get(clipped.size() - 1)) ) {
            clipped.remove(clipped.size() - 1);
        }
        return clipped;
    }
}

File: typescript/base/src/main/vsdk/toolkit/environment/geometry/geometricProcessing/polyhedralBoundedSolidOperators/fallbacks/_PolyhedralBoundedSolidFallbackGeometry.ts
import { Double } from "../../../../../../../java/lang/Double.js";
import { Vector3Dd } from "../../../../../common/linealAlgebra/Vector3Dd.js";
import type { PolyhedralBoundedSolid } from "../../../volume/polyhedralBoundedSolid/PolyhedralBoundedSolid.js";
import { _PolyhedralBoundedSolidOperator } from "../_PolyhedralBoundedSolidOperator.js";

/** Shared coordinate/profile primitives for structural boolean fallbacks. */
export class _PolyhedralBoundedSolidFallbackGeometry extends _PolyhedralBoundedSolidOperator {
    public static coordinate(point: Vector3Dd, axis: number): number {
        return axis === 0 ? point.x() : axis === 1 ? point.y() : point.z();
    }
    public static sameCoordinate(a: number, b: number): boolean {
        return Math.abs(a - b) <= this.numericContext.bigEpsilon();
    }
    public static boundsMatch(a: readonly number[] | null, b: readonly number[] | null): boolean {
        return (
            a !== null &&
            b !== null &&
            a.length >= 6 &&
            b.length >= 6 &&
            a.slice(0, 6).every((value, i) => this.sameCoordinate(value, b[i]!))
        );
    }
    public static addUniqueCoordinate(values: number[], value: number): void {
        if (values.some((existing) => this.sameCoordinate(existing, value))) return;
        values.push(value);
        // Java `Collections.sort` on `ArrayList<Double>` uses `Double.compareTo`.
        values.sort((a, b) => Double.compare(a, b));
    }
    public static uniqueVertexCoordinates(solid: PolyhedralBoundedSolid, axis: number): number[] {
        const values: number[] = [];
        for (let i = 0; i < solid.getVerticesList().size(); i++)
            this.addUniqueCoordinate(values, this.coordinate(solid.getVerticesList().get(i)!.position, axis));
        return values;
    }
    public static signedAreaOnYZ(profile: readonly Vector3Dd[]): number {
        let area = 0;
        for (let i = 0; i < profile.length; i++) {
            const a = profile[i]!,
                b = profile[(i + 1) % profile.length]!;
            area += a.y() * b.z() - b.y() * a.z();
        }
        return area * 0.5;
    }
    public static extractProfileAtX(solid: PolyhedralBoundedSolid, x: number): Vector3Dd[] | null {
        let best: Vector3Dd[] | null = null,
            bestArea = 0;
        for (let faceIndex = 0; faceIndex < solid.getPolygonsList().size(); faceIndex++) {
            const face = solid.getPolygonsList().get(faceIndex)!;
            for (let loopIndex = 0; loopIndex < face.boundariesList.size(); loopIndex++) {
                const loop = face.boundariesList.get(loopIndex)!;
                if (loop.halfEdgesList.size() < 3) continue;
                const profile: Vector3Dd[] = [];
                let onPlane = true;
                for (let i = 0; i < loop.halfEdgesList.size(); i++) {
                    const point = loop.halfEdgesList.get(i)!.startingVertex.position;
                    if (!this.sameCoordinate(point.x(), x)) {
                        onPlane = false;
                        break;
                    }
                    profile.push(new Vector3Dd(point));
                }
                const area = Math.abs(this.signedAreaOnYZ(profile));
                if (onPlane && area > bestArea) {
                    bestArea = area;
                    best = profile;
                }
            }
        }
        return best;
    }
    public static sameProfilePoint(a: Vector3Dd, b: Vector3Dd): boolean {
        return (
            this.sameCoordinate(a.x(), b.x()) && this.sameCoordinate(a.y(), b.y()) && this.sameCoordinate(a.z(), b.z())
        );
    }
    public static appendProfilePoint(profile: Vector3Dd[], point: Vector3Dd): void {
        if (profile.length > 0 && this.sameProfilePoint(profile[profile.length - 1]!, point)) return;
        profile.push(point);
    }
    public static projectProfilePoint(point: Vector3Dd, x: number, zCut: number): Vector3Dd {
        return new Vector3Dd(x, point.y(), this.sameCoordinate(point.z(), zCut) ? zCut : point.z());
    }
    public static intersectProfileSegmentAtZ(a: Vector3Dd, b: Vector3Dd, x: number, zCut: number): Vector3Dd {
        if (this.sameCoordinate(a.z(), b.z())) return new Vector3Dd(x, a.y(), zCut);
        const t = (zCut - a.z()) / (b.z() - a.z());
        return new Vector3Dd(x, a.y() + (b.y() - a.y()) * t, zCut);
    }
    public static clipProfileAboveZ(profile: readonly Vector3Dd[] | null, x: number, zCut: number): Vector3Dd[] {
        if (profile === null || profile.length < 3) return [];
        const clipped: Vector3Dd[] = [];
        let previous = profile[profile.length - 1]!,
            previousInside = previous.z() + this.numericContext.bigEpsilon() >= zCut;
        for (const current of profile) {
            const currentInside = current.z() + this.numericContext.bigEpsilon() >= zCut;
            if (currentInside) {
                if (!previousInside)
                    this.appendProfilePoint(clipped, this.intersectProfileSegmentAtZ(previous, current, x, zCut));
                this.appendProfilePoint(clipped, this.projectProfilePoint(current, x, zCut));
            } else if (previousInside)
                this.appendProfilePoint(clipped, this.intersectProfileSegmentAtZ(previous, current, x, zCut));
            previous = current;
            previousInside = currentInside;
        }
        if (clipped.length > 1 && this.sameProfilePoint(clipped[0]!, clipped[clipped.length - 1]!)) clipped.pop();
        return clipped;
    }
}
